components/txn_types/src/lock.rs

```diff
@@ -429,6 +430,14 @@ impl Lock {
 
         let raw_key = key.to_raw()?;
 
+        // Disable replica read for autocommit max ts read, to avoid breaking
+        // linearizability. See https://github.com/pingcap/tidb/issues/43583 for details.
+        if ts == TimeStamp::max() && is_replica_read {
+            return Err(Error::from(ErrorInner::KeyIsLocked(
+                lock.into_owned().into_lock_info(raw_key),
+            )));
+        }
+
         if ts == TimeStamp::max() && raw_key == lock.primary && !lock.use_async_commit {
             // When `ts == TimeStamp::max()` (which means to get latest committed version
             // for primary key), and current key is the primary key, we ignore
```

Comment:
What about move check here?

```rust
// Disable replica read for autocommit max ts read, to avoid breaking
// linearizability. See https://github.com/pingcap/tidb/issues/43583 for details.
if ts == TimeStamp::max() && raw_key == lock.primary && !lock.use_async_commit && !is_replica_read {
    // When `ts == TimeStamp::max()` (which means to get latest committed version
    // for primary key), and current key is the primary key, we ignore
    return Ok(());
}
```